### PR TITLE
HIVE-22826: ALTER TABLE RENAME COLUMN doesn't update list of bucketed…

### PR DESCRIPTION
**Pull request by ashish-kumar-sharma**
validate bucketed columns name in alter_table() and generate exception if wrong name is present in bucketed column.

### What changes were proposed in this pull request?

Compaction for tables where a bucketed column has been renamed fails since the list of bucketed columns in the StorageDescriptor doesn't get updated when the column is renamed, therefore we can't recreate the table correctly during compaction.

### Why are the changes needed?

Update bucket column name in AlterTableChangeColumnOperation along with column name.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test added

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java [hunk 1]
```diff
@@ -1554,4 +1555,30 @@ public static Partition createMetaPartitionObject(Table tbl, Map<String, String>
...
+  /**
+   * Validate bucket columns should belong to table columns.
+   * @param sd StorageDescriptor of given table
+   * @return true if bucket columns are empty or belong to table columns else false
+   */
+  public static boolean validateBucketColumns(StorageDescriptor sd) {
+    if (CollectionUtils.isNotEmpty(sd.getBucketCols())) {
+      List<String> columnNames =
+          getColumnNames(sd.getCols()).stream().map(String::toLowerCase).collect(Collectors.toList());
+      return columnNames.containsAll(sd.getBucketCols().stream().map(String::toLowerCase).collect(Collectors.toList()));
```
Comment (adesh-rao) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java:1568]: add a null check for columnNames? getColumnNames returns null too.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java [hunk 2]
```diff
@@ -63,6 +64,8 @@
 import org.datanucleus.api.jdo.JDOPersistenceManagerFactory;
 import org.junit.Assert;
 import org.junit.Before;
+import org.junit.experimental.categories.Category;
```
Comment (adesh-rao) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java:67]: required?

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java [hunk 3]
```diff
@@ -3428,4 +3432,44 @@ public void testUpdatePartitionStat_doesNotUpdateStats() throws Exception {
     m.invoke(hms, tbl, part, false, null);
     verify(wh, never()).getFileStatusesForLocation(part.getSd().getLocation());
   }
+
+  @Test(expected = InvalidOperationException.class)
```
Comment (adesh-rao) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java:3436]: Won't adding a qtest be better?

You are explicitly doing `tblUpdated.getSd().getBucketCols()` and updating them in the test. qtest will check end to end flow and you don't need to bother about getting Sd and updating the cols explicitly (Just a single alter table will do the trick).

Comment (adesh-rao) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java:3436]: Also, Is it a positive test case or negative? I see that there is an expected error but we are doing the positive validations too. We should break this into two, right?

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java (hunk 2 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java]: removed

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java (hunk 3 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java]: Divided the test in negative and positive

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/column/change/AlterTableChangeColumnOperation.java [hunk 4]
```diff
@@ -33,6 +35,7 @@
 import org.apache.hadoop.hive.ql.metadata.Partition;
 import org.apache.hadoop.hive.ql.metadata.Table;
 import org.apache.hadoop.hive.serde2.avro.AvroSerdeUtils;
+import org.apache.spark.util.CollectionsUtils;
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/column/change/AlterTableChangeColumnOperation.java]: Removed

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java [hunk 5]
```diff
@@ -130,6 +130,11 @@ public void alterTable(RawStore msdb, Warehouse wh, String catName, String dbnam
       throw new InvalidOperationException("Invalid column " + validate);
     }
 
+    // Validate bucketedColumns in new table
+    if (!MetaStoreServerUtils.validateBucketColumns(newt.getSd())) {
+      throw new InvalidOperationException("Bucket column doesn't match with any table columns");
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java:135]: Useful to add an error log with the column name which is missing from table columns list.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java [hunk 6]
```diff
@@ -1554,4 +1555,32 @@ public static Partition createMetaPartitionObject(Table tbl, Map<String, String>
...
     return tpart;
   }
+
+  /**
+   * Validate bucket columns should belong to table columns.
+   * @param sd StorageDescriptor of given table
+   * @return true if bucket columns are empty or belong to table columns else false
+   */
+  public static boolean validateBucketColumns(StorageDescriptor sd) {
+    List<String> columnNames = getColumnNames(sd.getCols());
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java:1565]: Check if CollectionUtils.isEmpty(sd.getBucketCols()) case even before getting table columns list and return true. It will avoid repeated checks below and also avoids unnecessary computes for non-bucketed tables.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java [hunk 7]
```diff
@@ -1554,4 +1555,32 @@ public static Partition createMetaPartitionObject(Table tbl, Map<String, String>
...
+    if(CollectionUtils.isNotEmpty(sd.getBucketCols()) &&  CollectionUtils.isNotEmpty(columnNames)){
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java:1566]: nit: Add space before "("

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java [hunk 8]
```diff
@@ -1554,4 +1555,32 @@ public static Partition createMetaPartitionObject(Table tbl, Map<String, String>
...
+  }
+
+  /**
+   * Generate column name list  from the fieldSchema list
+   * @param cols fieldSchema list
+   * @return column name list
+   */
+  public static List<String> getColumnNames(List<FieldSchema> cols) {
+    if (CollectionUtils.isNotEmpty(cols)) {
+      return cols.stream().map(FieldSchema::getName).collect(Collectors.toList());
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java:1582]: Will cols always have names in lower case?

**Review by sankarh — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java (hunk 8 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java]: Expected column name in lower case. But in order to be fail safe added toLowerCase() here also.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java (hunk 7 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java (hunk 6 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreServerUtils.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java (hunk 5 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java]: 1. Converted return type to List<String>.
2. Added Log.error() along with column name.
3. Added column to exception also.

File: ql/src/test/queries/clientpositive/alter_numbuckets_partitioned_table_h23.q [hunk 9]
```diff
@@ -52,6 +52,12 @@ alter table tst1_n1 clustered by (value) into 12 buckets;
 
 describe formatted tst1_n1;
 
+-- Test changing name of bucket column
+
+alter table tst1_n1 change key keys string;
+
+describe formatted tst1_n1;
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply ql/src/test/queries/clientpositive/alter_numbuckets_partitioned_table_h23.q:59]: Done

File: ql/src/test/queries/clientpositive/alter_bucketedtable_change_column.q [hunk 10]
```diff
@@ -0,0 +1,10 @@
+--! qt:dataset:src
+create table alter_bucket_change_col_t1(key string, value string) partitioned by (ds string) clustered by (key) into 10 buckets;
+
+describe formatted alter_bucket_change_col_t1;
+
+-- Test changing name of bucket column
+
+alter table alter_bucket_change_col_t1 change key keys string;
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply ql/src/test/queries/clientpositive/alter_bucketedtable_change_column.q]: added - "Serial_Num" which cover lower case, Upper case and also special char.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/column/change/AlterTableChangeColumnOperation.java [hunk 11]
```diff
@@ -72,6 +74,10 @@ protected void doAlteration(Table table, Partition partition) throws HiveExcepti
         if (desc.getNewColumnComment() != null) {
           oldColumn.setComment(desc.getNewColumnComment());
         }
+        if (CollectionUtils.isNotEmpty(sd.getBucketCols()) && sd.getBucketCols().contains(oldColumnName)) {
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/column/change/AlterTableChangeColumnOperation.java:76]: As per HIVE column contract it should be case in-sensitive. But it is not handled properly in  query planning of "alter table {tablename} change". So I have added toLowerCase() in query planning also.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/column/change/AlterTableChangeColumnOperation.java [hunk 12]
```diff
@@ -72,6 +74,10 @@ protected void doAlteration(Table table, Partition partition) throws HiveExcepti
...
+          sd.getBucketCols().remove(oldColumnName);
+          sd.getBucketCols().add(desc.getNewColumnName());
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/column/change/AlterTableChangeColumnOperation.java]: newColumnName is converted toLowerCase in query planning while populating "desc" but to be fail safe i have added toLowerCase() here also.

**Review by sankarh — APPROVED** [review]
+1, LGTM
